app/services/menu/loader.rb

```diff
@@ -34,7 +34,7 @@ def self.load
             menu.item :roles,            :caption => N_('Roles')
           end
           menu.divider
-          menu.item :bookmarks,          :caption => N_('Bookmarks')
+          menu.item :bookmarks,          :caption => N_('Bookmarks') if User.current && User.current.can?(:edit_bookmarks)
```

Comment:
a) This is completely in the wrong place, the loader shouldn't be checking permissions - it has no user context.  Permissions are checked per user using the routes/access_permissions list.
b) Changing the menu does nothing to fix the security issues in the ticket, the routes are still fully accessible and vulnerable.
c) The my_bookmarks scope serves its purpose to limit visibility of resources, but it's not used in the vulnerable controllers.  Adding taxonomies isn't really relevant to fixing this.